### PR TITLE
[MNG-5771] Document all classloaders used in Maven

### PR DESCRIPTION
**Pull request by kwin**
Clarify differences between Core and Build extensions

### CODE REVIEW

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 1]
```diff
@@ -30,55 +30,89 @@
...
 
  This is a description of the classloader hierarchy in Maven.
- 
+
+%{toc|section=1|fromDepth=2}
+
 * Overview
 
- * System Classloader
+ Maven uses the {{{https://codehaus-plexus.github.io/plexus-classworlds/}Plexus Classworlds}} classloading framework with which we
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:38]: "with which we
 create our" --> to create the

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 2]
```diff
@@ -30,55 +30,89 @@
...
+ create our classloader graph. If you look in your <<<$\{maven.home\}/boot>>> directory you will see a single JAR
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:39]: comma after directory

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 3]
```diff
@@ -30,55 +30,89 @@
...
+ which is the Classworlds JAR we use to boot the classloader graph. The Classworlds JAR is the only element of
+ the Java <<<CLASSPATH>>> and Classworlds then builds the other classloaders or realms in Classworlds terminology.
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:41]: "The Classworlds JAR is the only element of
 the Java <<<CLASSPATH>>> and Classworlds then builds the other classloaders or realms in Classworlds terminology." is awkward. perhaps something like

Classworlds uses this jar to bootstrap the other classloaders ("realms" in Classworlds terminology).

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 4]
```diff
@@ -30,55 +30,89 @@
...
- which is the Classworlds JAR we use to boot the classloader graph. The Classworlds JAR is the only element of
- the Java <<<CLASSPATH>>> and Classworlds then builds the other classloaders or realms in Classworlds terminology.
+* {Platform Classloader}
+
+ This is the classloader exposing all JRE classes.
+
+* {System Classloader}
  
- An Ant script like this will show the contents of the system classloader:
+ An Ant script like this will show the contents of the system/bootstrap classloader:
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:61]: why are we talking about ant? Please delete this.

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 5]
```diff
@@ -30,55 +30,89 @@
...
 +---+
 
+ It imports the platform classloader.
 
-* 2. {Core Classloader}
+* {Core Classloader}
  
- The second classloader down the graph contains the core requirements of Maven. More precisely, the core classloader
- has the libraries in <<<$\{maven.home\}/lib>>>. In general these are just Maven libraries, e.g. instances of
+ The second classloader down the graph contains the core requirements of Maven. <<It is only used by Maven internally but not by Plugins>>.
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:73]: delete only, lower case plugins

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 6]
```diff
@@ -30,55 +30,89 @@
...
 
+ It imports the platform classloader.
 
-* 2. {Core Classloader}
+* {Core Classloader}
  
- The second classloader down the graph contains the core requirements of Maven. More precisely, the core classloader
- has the libraries in <<<$\{maven.home\}/lib>>>. In general these are just Maven libraries, e.g. instances of
+ The second classloader down the graph contains the core requirements of Maven. <<It is only used by Maven internally but not by Plugins>>. 
+ The core classloader has the libraries in <<<$\{maven.home\}/lib>>>. In general these are just Maven libraries, e.g. instances of
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:74]: , e.g. -> . For example,

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 7]
```diff
@@ -30,55 +30,89 @@
...
  <<<{{{/ref/current/apidocs/org/apache/maven/project/MavenProject.html}MavenProject}}>>>
  belong to this classloader.
- We hope to further separate these in the future to just be Maven APIs and have 
- the implementations selected at runtime as required by the system.
  
  You can add elements to this classloader by {{{/ref/current/maven-model/maven.html#class_extension}extensions}}. 
  These are loaded into the same place as <<<$\{maven.home\}/lib>>> and hence are available
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:79]: place is unclear. Is this a classpath?

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 8]
```diff
@@ -30,55 +30,89 @@
...
  belong to this classloader.
- We hope to further separate these in the future to just be Maven APIs and have 
- the implementations selected at runtime as required by the system.
  
  You can add elements to this classloader by {{{/ref/current/maven-model/maven.html#class_extension}extensions}}. 
  These are loaded into the same place as <<<$\{maven.home\}/lib>>> and hence are available 
- to the Maven core and all plugins for the current project and subsequent projects (in future, we
- plan to remove it from subsequent projects).
+ to the Maven core and all plugins for the current project (through the API classloader).
+ More information available in {{{./guide-using-extensions.html}Core Extension}}.
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:81]: is available

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 9]
```diff
@@ -30,55 +30,89 @@
...
+
+* {API Classloader}
+
+ This exposes a filtered view of the Core Classloader by only exposing the exported packages from all Core Extensions.
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:85]: This --> The API classloader

only exposing the --> exposing only the

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 10]
```diff
@@ -30,55 +30,89 @@
...
+
+* {Build Extension Classloaders}
 
-  
-* 3. {Plugin Classloaders}
+[../../buildExtensionClassRealm.svg] Build Extension Class Realm
 
- After that, each plugin has its own classloader that is a child of Maven's core classloader. 
- The classes in this classloader are taken from the dependencies in the plugin's dependency list.
+ For every plugin which is marked with <<<\<extensions\>true\</extensions\>>>> or a build extension listed in the according section of the POM there is
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:94]: comma after POM

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 11]
```diff
@@ -30,55 +30,89 @@
...
+ a dedicated classloader. Those are isolated (i.e. one build extension does not have access to other build extensions).
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:95]: Those are isolated (i.e. -->

Those are isolated. That is,

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 12]
```diff
@@ -30,55 +30,89 @@
...
+ It imports everything from API classloader.
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:96]: the API

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 13]
```diff
@@ -30,55 +30,89 @@
...
+ All JSR 330 or Plexus components declared in the underlying JAR are registered as components in the global Plexus container
+ while creating the classloader.
+ In addition all component references in the plugin descriptor are properly wired from the underlying Plexus container.
+ Build extensions have limited effect as they are loaded late.
+
+* {Project Classloaders}
+
+ There is one project classloader per Maven project (identified through its GAV).
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:104]: GAV --> coordinates

File: content/apt/guides/mini/guide-maven-classloading.apt [hunk 14]
```diff
@@ -118,7 +152,7 @@ Guide to Maven Classloading
  When a build plugin is executed, the thread's context classloader is set to the plugin classloader.
 
 
-* 4. {Custom Classloaders}
+* {Custom Classloaders}
  
  Plugins are free to create further classloaders on their discretion. For example, a plugin might want to create a
```
Comment (elharo) [inline content/apt/guides/mini/guide-maven-classloading.apt:157]: delete "on their discretion"

File: content/apt/guides/mini/guide-using-extensions.apt [hunk 15]
```diff
@@ -28,7 +28,8 @@
 
 Using Extensions
 
- Extensions are a way to add classes to {{{./guide-maven-classloading.html#Core_Classloader}Core Classloader}}.
+ Extensions are a way to add classes to either the {{{./guide-maven-classloading.html#Core_Classloader}Core Classloader}} (Core Extensions) 
+ or the {{{./guide-maven-classloading.html#Project_Classloaders}Project Classloader}} (Build Extensions).
  This is necessary for adjusting Maven in a way that affects more than just one plug-in.
 
  The mechanism allows to either replace default {{{https://www.eclipse.org/sisu/}Sisu components}} with custom ones or add
```
Comment (elharo) [inline content/apt/guides/mini/guide-using-extensions.apt:35]: allows who?

File: content/apt/guides/mini/guide-using-extensions.apt [hunk 16]
```diff
@@ -40,10 +41,12 @@ Using Extensions
 
 * Loading Extensions
 
- There are different means of loading extensions. They can be distinguished into <<early>> and <<late>> loading behaviour. Some extensions require early loading.
- Please refer to the extension's documentation whether early loading is required.
+ There are different means of loading extensions depending on the type. There are <core extensions> which are loaded <<early>>
+ and build extensions which are loaded <<late>>.
+ Some extensions require early loading as they fundamentally change Maven behaviour.
+ Please refer to the extension's documentation whether they provide a core or a build extension.
```
Comment (elharo) [inline content/apt/guides/mini/guide-using-extensions.apt:47]: "Please refer to the extension's documentation whether they provide a core or a build extension." --> An extension's documentation should indicate whether it provide a core or a build extension.

File: src/plantuml/buildExtensionClassRealm.puml [hunk 17]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (elharo) [inline src/plantuml/buildExtensionClassRealm.puml:1]: What are these new files doing here? Maybe drawing some sort of diagram or something?

**Review by elharo — DISMISSED** [review]
Thanks for cleaning this up. Looking at it now, this was indeed very unclear.

**Comment by cstamas** [comment]
Starter comment: could we just convert any exotic formatted document to something supported by modern tools when we touch them? Like Markdown or ADoc (depdending on complexity and requirements of documentation)?
I really dont understand why on earth we keep coming up with our own "standards" that just make our lives harder than should :smile:

**Comment by kwin** [comment]
> Starter comment: could we just convert any exotic formatted document to something supported by modern tools when we touch them? Like Markdown or ADoc (depdending on complexity and requirements of documentation)? I really dont understand why on earth we keep coming up with our own "standards" that just make our lives harder than should 😄

I don't think this should be included in PRs which change something else. Otherwise the diff is no longer visible. It needs to be separately done either before or after. Also I miss the automated conversion from `APT` to `Markdown` in https://github.com/apache/maven-doxia-converter....

**Comment by cstamas** [comment]
Wow, yet another Maven Component that depends on plexus-container-default... :beer:

**Comment by elharo** [comment]
It might be worthwhile to write a general purpose .apt to markdown converter. However that should not be part of this PR. As to why we have these exotic special purpose things, a lot of it is is simply that this code goes back 20 years and no such de facto standard tool existed back then.

File: src/plantuml/buildExtensionClassRealm.puml (hunk 17 above)
Comment (kwin) [reply src/plantuml/buildExtensionClassRealm.puml:1]: Yes, exactly, they have this format: https://plantuml.com/class-diagram and are converted during build to SVG. Also they are referenced in modified APT documents e.g. in https://github.com/apache/maven-site/pull/338/files#diff-3a01086d28efd2b3cb838d37d2efcafa96e65044d0b09a6826b6a5ca3b54a45aR92.

**Review by michael-o — APPROVED** [review]

**Comment by kwin** [comment]
@cstamas Any further remarks before I go ahead and merge?

**Review by cstamas — APPROVED** [review]

**Pull request merged**